Tools/ML/ONNX/model.cxx

```diff
@@ -0,0 +1,213 @@
+// All rights not expressly granted are reserved.
+//
+// This software is distributed under the terms of the GNU General Public
+// License v3 (GPL Version 3), copied verbatim in the file "COPYING".
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+///
+/// \file     model.cxx
+///
+/// \author   Christian Sonnabend <christian.sonnabend@cern.ch>
+///
+/// \brief    A general-purpose class with functions for ONNX model applications
+///
+
+// ONNX includes
+#include "Tools/ML/ONNX/model.h"
+
+namespace o2
+{
+
+namespace ml
+{
+
+std::string OnnxModel::printShape(const std::vector<int64_t>& v)
+{
+  std::stringstream ss("");
+  for (size_t i = 0; i < v.size() - 1; i++)
+    ss << v[i] << "x";
+  ss << v[v.size() - 1];
+  return ss.str();
+}
+
+OnnxModel& OnnxModel::operator=(OnnxModel& inst)
+{
+
+  mEnv = inst.mEnv;
+  mSession = inst.mSession;
+
+  mInputNames = inst.mInputNames;
+  mInputShapes = inst.mInputShapes;
+  mOutputNames = inst.mOutputNames;
+  mOutputShapes = inst.mOutputShapes;
+
+  modelPath = inst.modelPath;
+  ccdbUrl = inst.ccdbUrl;
+  activeThreads = inst.activeThreads;
+  valid_from = inst.valid_from;
+  valid_until = inst.valid_until;
+
+  LOG(debug) << "Model copied!";
+
+  return *this;
+}
+
+bool OnnxModel::checkHyperloop()
+{
+  /// Testing hyperloop core settings
+  const char* alien_cores = gSystem->Getenv("ALIEN_JDL_CPUCORES");
+  bool alien_cores_found = (alien_cores != NULL);
+  if (alien_cores_found) {
+    LOGP(info, "Hyperloop test/Grid job detected! Number of cores = {}. Setting threads anyway to 1.", alien_cores);
+    activeThreads = 1;
+  } else {
+    LOGP(info, "Not running on Hyperloop.");
+  }
+
+  return alien_cores_found;
+}
+
+OnnxModel::OnnxModel(std::string path)
+{
+
+  LOG(info) << "--- ONNX-ML model ---";
+
+  modelPath = path;
+  mEnv = std::make_shared<Ort::Env>(ORT_LOGGING_LEVEL_WARNING, "onnx-model");
+  mSession.reset(new Ort::Experimental::Session{*mEnv, path, sessionOptions});
+
+  mInputNames = mSession->GetInputNames();
+  mInputShapes = mSession->GetInputShapes();
+  mOutputNames = mSession->GetOutputNames();
+  mOutputShapes = mSession->GetOutputShapes();
+
+  LOG(info) << "Input Nodes:";
+  for (size_t i = 0; i < mInputNames.size(); i++) {
+    LOG(info) << "\t" << mInputNames[i] << " : " << printShape(mInputShapes[i]);
+  }
+
+  LOG(info) << "Output Nodes:";
+  for (size_t i = 0; i < mOutputNames.size(); i++) {
+    LOG(info) << "\t" << mOutputNames[i] << " : " << printShape(mOutputShapes[i]);
+  }
+
+  LOG(info) << "--- Model initialized! ---";
+
+  if (checkHyperloop()) {
+    sessionOptions.SetIntraOpNumThreads(activeThreads);
+  }
+}
+
+bool OnnxModel::fetchFromCCDB(std::string path_from, int64_t ccdbTimestamp, std::string path_to = "model.onnx")
```

Comment:
You can avoid copying the code if you call `downloadToFile()` inside `fetchFromCCDB()`